### PR TITLE
NewObjectToFactoryCreateRector

### LINKED ISSUE #981: Idea - NewToFactoryRector
Please @TomasVotruba feedback needed:

Before:
```php
public function xyz() {
    $a = new MyClass();
}
```

After:
```php
/** @var MyClassFactory */
private $myClassFactory;

public function xyz() {
    $a = $this->myClassFactory->createMyClass();
}
```

I am not sure about the config. We need to configure:  
1) Instance class
2) Factory class
3) Property visibility? Or set it private by default?
4) Property name

Can it works like this:
```yaml
services:
  MyFactoryRectorName:
    MyClass: # 1) instance class to get replaced with factory call
      factoryClass: MyClassFactory # 2) factory class
      factoryPropertyName: $myClassFactory # 3) property name
      factoryPropertyVisibility: private # 4) property visibility
```

What do you think? Will it be possible to access these config options in rector class?

### PR DESCRIPTION
**Pull request by JanMikes**
Closes #981

### CODE REVIEW

**Comment by JanMikes** [comment]
@TomasVotruba please could you check my code?

I am getting weird error with this config:
```
Symfony\Component\DependencyInjection\Exception\InvalidArgumentException : The configuration key "Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClass" is unsupported for definition "Rector\Rector\Architecture\Factory\NewObjectToFactoryCreateRector" in "/Users/janmikes/Sites/rector/tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/config.yml". Allowed configuration keys are "alias", "parent", "class", "shared", "synthetic", "lazy", "public", "abstract", "deprecated", "factory", "file", "arguments", "properties", "configurator", "calls", "tags", "decorates", "decoration_inner_name", "decoration_priority", "autowire", "autoconfigure", "bind".
 /Users/janmikes/Sites/rector/vendor/symfony/dependency-injection/Loader/YamlFileLoader.php:813
 /Users/janmikes/Sites/rector/vendor/symfony/dependency-injection/Loader/YamlFileLoader.php:341
 /Users/janmikes/Sites/rector/vendor/symfony/dependency-injection/Loader/YamlFileLoader.php:229
 /Users/janmikes/Sites/rector/vendor/symplify/package-builder/src/Reflection/PrivatesCaller.php:23
 /Users/janmikes/Sites/rector/vendor/symplify/package-builder/src/Yaml/FileLoader/AbstractParameterMergingYamlFileLoader.php:106
 /Users/janmikes/Sites/rector/vendor/symfony/config/Loader/DelegatingLoader.php:40
 /Users/janmikes/Sites/rector/src/DependencyInjection/RectorKernel.php:50
 /Users/janmikes/Sites/rector/vendor/symfony/http-kernel/Kernel.php:653
 /Users/janmikes/Sites/rector/vendor/symfony/http-kernel/Kernel.php:543
 /Users/janmikes/Sites/rector/vendor/symfony/http-kernel/Kernel.php:133
 /Users/janmikes/Sites/rector/src/DependencyInjection/ContainerFactory.php:26
 /Users/janmikes/Sites/rector/src/Testing/PHPUnit/AbstractRectorTestCase.php:142
 /Users/janmikes/Sites/rector/src/Testing/PHPUnit/AbstractRectorTestCase.php:64
```

**Comment by JanMikes** [comment]
Nevermind, i fixed that. Not using `provideConfig()` but `getRectorConfiguration()` did the trick 👍

File: tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/config.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+services:
+  Rector\Rector\Architecture\Factory\NewObjectToFactoryCreateRector:
+    Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClass: []
```
Comment (TomasVotruba) [inline tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/config.yml]: Please use 4-spacing as everywhere here. It shold be loaded `.editorconfig`. Do you have it installed in PHPStorm?

File: tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/NewObjectToFactoryCreateRectorTest.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<?php declare(strict_types=1);
+
+namespace Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector;
+
+use Rector\Testing\PHPUnit\AbstractRectorTestCase;
+
+/**
+ * @covers \Rector\Rector\Architecture\Factory\NewObjectToFactoryCreateRector
+ * @covers \Rector\Rector\Architecture\DependencyInjection\ReplaceVariableByPropertyFetchRector
```
Comment (TomasVotruba) [inline tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/NewObjectToFactoryCreateRectorTest.php]: This is unneded step. Use just single Rector to do this, so people won't forget

File: tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/config.yml (hunk 1 above)
Comment (JanMikes) [reply tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/config.yml]: No i did not, thanks for perfect tip 👍

File: tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/config.yml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+services:
+  Rector\Rector\Architecture\Factory\NewObjectToFactoryCreateRector:
+    arguments:
```
Comment (JanMikes) [inline tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/config.yml]: @TomasVotruba well, this works, but please how would the config look like, if i wanted to use more objects and more factories, because `arguments` key is unique for this service. Sorry for struggling with this basic symfony/yaml feature

Comment (TomasVotruba) [reply tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/config.yml]: That's part of learning Symfony.

It took me while to find this feature, it's here: https://symfony.com/blog/new-in-symfony-3-3-simpler-service-configuration#short-syntax-for-service-configuration

File: tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/NewObjectToFactoryCreateRectorTest.php (hunk 2 above)
Comment (TomasVotruba) [reply tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/NewObjectToFactoryCreateRectorTest.php]: Yea, the goal of this annotaiton is to get to tested class quickly. Just control + click in PHPStorm.

In 99 % cases it's in `getRectorClass()` so it's not needed here.

**Comment by TomasVotruba** [comment]
```php
<?php

final class NewObjectToFactoryCreateRector extends AbstractRector
{
    /**
     * @var string[][]
     */
    private $factoryWithMethodByObject;

    /**
     * @param string[][] $factoryWithMethodByObject
     */
    public function __construct(array $factoryWithMethodByObject)
    {
        $this->factoryWithMethodByObject = $factoryWithMethodByObject;
    }
    
    /**
     * @return string[]
     */
    public function getNodeTypes(): array
    {
        return [New_::class];
    }

    /**
     * @param New_ $node
     */
    public function refactor(Node $node): ?Node
    {
        foreach ($this->factoryWithMethodByObject as $type => $factoryWithMethod) {
            if (! $this->isType($node, $type)) {
                continue;
            }

            // reactor to MethodCall
            // add property
        }

        return $node;
    }
}
```

This will be then configured in config:


```yaml
# rector.yml
services:
	NewObjectToFactoryCreateRector: # ← rector class
		# configuration array that will be passed as $factoryWithMethodByObject
		FirstObject: ['FirstFactory', 'method']
		SecondObject: ['SecondFactory', 'method']
```

↓ full syntax without magic. This would people have to write manually and be careful about typos in "factoryWithMethodByObject", which can be really annoying.

```yaml
services:
	NewObjectToFactoryCreateRector: # ← rector class
	    arguments:
	    	$factoryWithMethodByObject:
				FirstObject: ['FirstFactory', 'method']
				SecondObject: ['SecondFactory', 'method']
```

**Comment by TomasVotruba** [comment]
Could you push it? Now there is no array in constructor to put it to

**Comment by TomasVotruba** [comment]
It works as expected. Here is the change in code: https://github.com/rectorphp/rector/compare/new-object-to-factory-method-rector...quick-example?expand=1

**Comment by JanMikes** [comment]
Btw i did not test/cover edge cases - property already exists, factory already exists in different property in class etc.

**Comment by TomasVotruba** [comment]
@JanMikes I see some commits already in master. Could you rebase on it?

image

**Comment by TomasVotruba** [comment]
Feature works, nice :+1: 

Testing edge cases would be nice to have, but not neccesary. It's up to you.
I think we'll find them when using on Entrydo.

Also there is 1 fail on static analysis, please fix it:

https://travis-ci.org/rectorphp/rector/jobs/483997841#L661

**Comment by JanMikes** [comment]
Thank you, i fixed that and i will as well cover the edge cases, while i am already in it 😉 

I think that history must be mistaken, i have just checked it and there are no commits in master yet -> i rebased this branch on master. Last my commit i can see in master is  `31a5e9bcd2c44a1553fd7f17f4d4c563c138af3b`

**Comment by JanMikes** [comment]
This edge case covered by: https://github.com/rectorphp/rector/blob/43de77013c5ca172fb53a87cb35ec13e2391c289/tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/Fixture/fixture2.php.inc

Now it is definitely ready to merge 😉

**Comment by TomasVotruba** [comment]
I'm on it :)

**Comment by TomasVotruba** [comment]
Btw, no need to rebase after each commit, just once in the end.
Now I don't know what's new since the last review and have to read whole code.

File: tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/Fixture/fixture.php.inc [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Fixture;
+
+use Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClass;
+
+final class SomeController
+{
+    public function default()
+    {
+        new MyClass('abcd');
```
Comment (TomasVotruba) [inline tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/Fixture/fixture.php.inc]: Could add a new line under this one with assign case?

```php
$myClass = new MyClass('abcd');
```

File: tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/Fixture/fixture2.php.inc [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    public function __construct(\Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClassFactory $mySomeFactory)
+    {
+        $this->mySomeFactory = $mySomeFactory;
+    }
+    public function default()
+    {
+        $this->mySomeFactory->create('abcd');
+    }
+}
+?>
```
Comment (TomasVotruba) [inline tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/Fixture/fixture2.php.inc]: Awesome :cupid:

File: tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/config.yml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+services:
+    Rector\Rector\Architecture\Factory\NewObjectToFactoryCreateRector:
+        Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClass:
+            - Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClassFactory
+            - create
```
Comment (TomasVotruba) [inline tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/config.yml]: What about adding `class` and `method` keys, so it's more obvious this is not n-list?

```diff
 services:
     Rector\Rector\Architecture\Factory\NewObjectToFactoryCreateRector:
         Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClass:
-            - Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClassFactory
-            - create
+            class: Rector\Tests\Rector\Architecture\Factory\NewObjectToFactoryCreateRector\Source\MyClassFactory
+            method create
```

File: src/Rector/Architecture/Factory/NewObjectToFactoryCreateRector.php [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	 */
+	private $myClassFactory;
+
+	public function example() {
+		$this->myClassFactory->create($argument);
+	}
+}
+CODE_SAMPLE
+                ,
+                ['MyClass' => ['MyClassFactory', 'create']]
```
Comment (TomasVotruba) [inline src/Rector/Architecture/Factory/NewObjectToFactoryCreateRector.php]: :+1:

File: tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/Fixture/fixture.php.inc (hunk 4 above)
Comment (JanMikes) [reply tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/Fixture/fixture.php.inc]: Good catch! This causes error and i have to figure out and fix it 😉 👍 

Anyways, it is terrible for debugging with xdebug, i can not see what exact function call caused it and overall i have no idea 😄 

screenshot 2019-01-24 22 58 10

Comment (JanMikes) [reply tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/Fixture/fixture.php.inc]: After quick research i found out this is the problem:

```
$this->addNodeAfterNode($methodCallNode, $node);
$this->removeNode($node);
```

Because `removeNode()` only works if it is `new xx()` without variable assign.

I will figure out and dig more and let you know 😉 im looking for elegant way of replacing New with MethodCall node.

Comment (JanMikes) [reply tests/Rector/Architecture/Factory/NewObjectToFactoryCreateRector/Fixture/fixture.php.inc]: Simple `return $methodCallNode;` did the trick 🎉

**Comment by TomasVotruba** [comment]
Great! Ready to merge for me :+1:

**Comment by TomasVotruba** [comment]
Kaboom :rocket:

**Comment by JanMikes** [comment]
Looking forward to get this tagged 😉 for now i will use dev-master dependency.
